File: frontend/src/types/blueprint.ts
// ============================================================================
// BLUEPRINT V2 TYPES - Dynamic UI Blocks System
// ============================================================================

// --- Data Layer ---

export interface ColumnSpec {
   name: string;
   type: 'uuid' | 'text' | 'int' | 'float' | 'bool' | 'date' | 'timestamptz' | 'jsonb';
   required?: boolean;
   default?: any;
   unique?: boolean;
   indexed?: boolean;
}

export interface TableSpec {
   name: string;
   primaryKey?: string;
   columns: ColumnSpec[];
}

export interface RelationshipSpec {
   type: 'many_to_one' | 'one_to_many';
   fromTable: string;
   fromColumn: string;
   toTable: string;
   toColumn: string;
   lookupLabelColumn?: string;
}

export interface DataSpec {
   tables: TableSpec[];
   relationships?: RelationshipSpec[];
}

// --- Security Layer ---

export interface PermissionRule {
   role: string;
   resource: string;
   actions: Record<string, boolean>;
}

export interface RowFilterRule {
   role: string;
   resource: string;
   filter: any;
}

export interface SecuritySpec {
   roles: string[];
   permissions: PermissionRule[];
   rowFilters?: RowFilterRule[];
}

// --- UI Blocks System ---

export interface FilterSpec {
   field: string;
   operator: 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte' | 'like' | 'in' | 'is_null' | 'is_not_null';
   value: any;
}

export interface OrderSpec {
   field: string;
   direction: 'asc' | 'desc';
}

export interface DataSourceSpec {
   table: string;
   filters?: FilterSpec[];
   orderBy?: OrderSpec[];
   limit?: number;
   include?: string[];
   realtime?: boolean;
}

export interface ActionConfig {
   trigger: string;
   action: string;
   config?: Record<string, any>;
   condition?: string;
}

export interface VisibilityRule {
   condition: string;
}

// --- Block Types ---

export type BlockType =
   | 'table'
   | 'form'
   | 'detail'
   | 'stat-card'
   | 'chart'
   | 'kanban'
   | 'calendar'
   | 'timeline'
   | 'chat'
   | 'gallery'
   | 'tree'
   | 'file-list'
   | 'container'
   | 'tabs'
   | 'split-view'
   | 'button'
   | 'search'
   | 'filter-bar'
   | 'markdown';

export interface BlockSpec {
   id: string;
   type: BlockType | string;
   dataSource?: DataSourceSpec;
   props: Record<string, any>;
   actions?: ActionConfig[];
   gridArea?: string;
   className?: string;
   visibility?: VisibilityRule;
   children?: BlockSpec[];
}

// --- Block Props Types ---

export interface TableColumnDef {
   field: string;
   label: string;
   type?: 'text' | 'number' | 'date' | 'datetime' | 'badge' | 'boolean' | 'image' | 'link';
   sortable?: boolean;
   width?: string;
}

export interface TableProps {
   columns: TableColumnDef[];
   allowSearch?: boolean;
   allowFilter?: boolean;
   allowSort?: boolean;
   allowSelect?: boolean;
   allowExport?: boolean;
   pagination?: { pageSize: number };
   rowActions?: ('view' | 'edit' | 'delete' | ActionConfig)[];
   emptyMessage?: string;
}

export interface FormFieldDef {
   name: string;
   label: string;
   type: 'text' | 'textarea' | 'number' | 'select' | 'checkbox' | 'date' | 'datetime' | 'relation' | 'file' | 'color' | 'email' | 'password' | 'url';
   required?: boolean;
   options?: string[] | { value: string; label: string }[];
   placeholder?: string;
   defaultValue?: any;
   table?: string; // For relation fields
}

export interface FormProps {
   mode: 'create' | 'edit';
   fields: FormFieldDef[];
   submitLabel?: string;
   cancelLabel?: string;
}

export interface DetailFieldDef {
   name: string;
   label: string;
   type?: 'text' | 'date' | 'datetime' | 'badge' | 'boolean' | 'image' | 'link';
}

export interface DetailProps {
   fields: DetailFieldDef[];
   layout?: 'vertical' | 'horizontal' | 'grid';
}

export interface StatCardProps {
   title: string;
   value?: string | number;
   valueField?: string;
   icon?: string;
   trend?: number;
   trendDirection?: 'up' | 'down';
   color?: string;
   query?: string;
}

export interface ChartProps {
   chartType: 'bar' | 'line' | 'pie' | 'donut' | 'area' | 'scatter';
   xField?: string;
   yField?: string;
   groupField?: string;
   aggregation?: 'count' | 'sum' | 'avg' | 'min' | 'max';
   colors?: string[];
   showLegend?: boolean;
   showGrid?: boolean;
   animate?: boolean;
}

export interface KanbanColumnDef {
   value: string;
   label: string;
   color?: string;
   limit?: number;
}

export interface KanbanCardConfig {
   titleField: string;
   descriptionField?: string;
   metaFields?: string[];
   badgeField?: string;
   badgeColors?: Record<string, string>;
   avatarField?: string;
}

export interface KanbanProps {
   groupByField: string;
   columns: KanbanColumnDef[];
   card: KanbanCardConfig;
   allowDragDrop?: boolean;
   allowCreate?: boolean;
   allowCollapse?: boolean;
}

export interface CalendarProps {
   startField: string;
   endField: string;
   titleField: string;
   colorField?: string;
   colors?: Record<string, string>;
   views?: ('month' | 'week' | 'day' | 'agenda')[];
   defaultView?: string;
   allowCreate?: boolean;
   allowDrag?: boolean;
   allowResize?: boolean;
   minTime?: string;
   maxTime?: string;
}

export interface TimelineProps {
   dateField: string;
   titleField: string;
   descriptionField?: string;
   groupBy?: 'day' | 'week' | 'month';
   showTime?: boolean;
   iconField?: string;
   colorField?: string;
}

export interface ChatProps {
   messageField: string;
   senderNameField: string;
   senderAvatarField?: string;
   timestampField: string;
   allowReply?: boolean;
   allowReactions?: boolean;
   allowAttachments?: boolean;
   allowEdit?: boolean;
   allowDelete?: boolean;
   realtime?: boolean;
}

export interface GalleryProps {
   imageField: string;
   titleField?: string;
   descriptionField?: string;
   columns?: number;
   aspectRatio?: '1:1' | '16:9' | '4:3' | '3:2';
   allowLightbox?: boolean;
}

export interface TreeProps {
   titleField: string;
   parentField?: string;
   iconField?: string;
   expandable?: boolean;
   selectable?: boolean;
}

// --- Layout System ---

export interface LayoutConfig {
   type: 'single' | 'split' | 'grid' | 'tabs';
   config?: {
      // For split
      direction?: 'horizontal' | 'vertical';
      sizes?: number[];
      resizable?: boolean;
      // For grid
      columns?: number;
      rows?: string;
      gap?: string;
      // For tabs
      position?: 'top' | 'left' | 'bottom';
      tabs?: { id: string; label: string; icon?: string }[];
      // For single
      maxWidth?: string;
      padding?: string;
   };
}

// --- Page & Modal Definitions ---

export interface PageSpec {
   id: string;
   route: string;
   title: string;
   icon?: string;
   layout?: LayoutConfig;
   blocks: BlockSpec[];
   variables?: Record<string, any>;
}

export interface ModalSpec {
   id: string;
   title: string;
   size?: 'small' | 'medium' | 'large' | 'fullscreen';
   blocks: BlockSpec[];
}

export interface NavItem {
   name: string;
   label: string;
   icon?: string;
   route?: string;
   children?: NavItem[];
}

export interface GlobalAction {
   id: string;
   label: string;
   icon?: string;
   action: string;
   config: Record<string, any>;
}

// --- Theme ---

export interface ThemeSpec {
   primaryColor?: string;
   mode?: 'light' | 'dark';
   fontFamily?: string;
}

// --- App Info ---

export interface AppInfo {
   name: string;
   slug: string;
   description?: string;
   theme?: ThemeSpec;
}

// --- UI Spec ---

export interface UISpec {
   navigation: NavItem[];
   pages: PageSpec[];
   modals?: ModalSpec[];
   globalActions?: GlobalAction[];
}

// --- Blueprint V2 (Main Type) ---

export interface BlueprintV2 {
   version: 2;
   app: AppInfo;
   data: DataSpec;
   security: SecuritySpec;
   ui: UISpec;
}

// --- Legacy V1 Support ---

export interface ResourceSpec {
   name: string;
   table: string;
   label: string;
   views?: Record<string, boolean>;
   list?: { columns: string[] };
   forms?: {
      createFields?: { name: string; widget?: string; label?: string; options?: any[] }[];
      editFields?: { name: string; widget?: string; label?: string; options?: any[] }[];
   };
}

export interface UISpecV1 {
   navigation: NavItem[];
   resources: ResourceSpec[];
   pages?: any[];
}

export interface BlueprintV1 {
   version: 1;
   app: AppInfo;
   data: DataSpec;
   security: SecuritySpec;
   ui: UISpecV1;
}

// Union type for both versions
export type Blueprint = BlueprintV1 | BlueprintV2;

// Type guard to check blueprint version
export function isBlueprintV2(blueprint: Blueprint): blueprint is BlueprintV2 {
   return blueprint.version === 2 && 'pages' in blueprint.ui;
}

export function isBlueprintV1(blueprint: Blueprint): blueprint is BlueprintV1 {
   return blueprint.version === 1 || 'resources' in blueprint.ui;
}

File: backend/app/schemas/blueprint.py
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal, Union
from app.models.blueprint import ValidationStatus


# ============================================================================
# BLUEPRINT V3 SCHEMA - Full-Stack Generation with Amplication
# ============================================================================

# --- Data Layer ---

class ColumnSpec(BaseModel):
   name: str
   type: Literal["uuid", "text", "int", "float", "bool", "date", "timestamptz", "jsonb"]
   required: bool = False
   default: Optional[Any] = None
   unique: bool = False
   indexed: bool = False


class TableSpec(BaseModel):
   name: str  # PascalCase for V3 (e.g., "Task", "Project")
   displayName: Optional[str] = None
   primaryKey: str = "id"
   columns: List[ColumnSpec]


class RelationshipSpec(BaseModel):
   name: str  # Relationship name (e.g., "project", "tasks")
   type: Literal["many_to_one", "one_to_many"]
   fromTable: str
   toTable: str
   fromColumn: Optional[str] = None  # Optional, auto-generated if not provided
   toColumn: Optional[str] = None
   lookupLabelColumn: Optional[str] = None


class DataSpec(BaseModel):
   tables: List[TableSpec]
   relationships: Optional[List[RelationshipSpec]] = []


# --- Security Layer V3 (Entity-Action based) ---

class RoleSpec(BaseModel):
   name: str
   displayName: Optional[str] = None


class EntityPermission(BaseModel):
   role: str
   entity: str
   actions: Dict[str, bool]  # create, read, update, delete


class SecuritySpecV3(BaseModel):
   roles: List[RoleSpec]
   permissions: List[EntityPermission]


# --- Backend Configuration (V3 only) ---

class AuthConfig(BaseModel):
   provider: Literal["jwt", "basic", "none"] = "jwt"


class BackendSettings(BaseModel):
   generateREST: bool = True
   generateGraphQL: bool = False
   generateSwagger: bool = True


class BackendConfig(BaseModel):
   generator: Literal["amplication"] = "amplication"
   settings: BackendSettings = BackendSettings()
   auth: AuthConfig = AuthConfig()


# --- UI Blocks System ---

class FilterSpec(BaseModel):
   field: str
   operator: Literal["eq", "neq", "gt", "gte", "lt", "lte", "like", "in", "is_null", "is_not_null"]
   value: Any


class OrderSpec(BaseModel):
   field: str
   direction: Literal["asc", "desc"] = "asc"


class DataSourceSpec(BaseModel):
   entity: str  # V3 uses "entity" instead of "table"
   filters: Optional[List[FilterSpec]] = []
   orderBy: Optional[List[OrderSpec]] = []
   limit: Optional[int] = None
   include: Optional[List[str]] = []  # Relations to include
   realtime: Optional[bool] = False


class ActionConfig(BaseModel):
   trigger: str  # click, submit, change, cardClick, cardMove, etc.
   action: str   # navigate, openModal, closeModal, createRecord, updateRecord, deleteRecord, apiCall, etc.
   config: Optional[Dict[str, Any]] = {}
   condition: Optional[str] = None


class VisibilityRule(BaseModel):
   condition: str  # Expression like "{{$user.role}} === 'admin'"


# --- Block Specifications ---

class BlockSpec(BaseModel):
   id: str
   type: str  # TABLE, FORM, KANBAN, CALENDAR, CHART, STAT_CARD, etc.
   dataSource: Optional[DataSourceSpec] = None
   props: Dict[str, Any] = {}
   actions: Optional[List[ActionConfig]] = []
   gridArea: Optional[str] = None  # CSS grid placement
   className: Optional[str] = None
   visibility: Optional[VisibilityRule] = None
   children: Optional[List["BlockSpec"]] = []


# --- Layout System ---

class LayoutConfig(BaseModel):
   type: Literal["single", "split", "grid", "tabs"] = "single"
   config: Optional[Dict[str, Any]] = {}


class TabConfig(BaseModel):
   id: str
   label: str
   icon: Optional[str] = None


# --- Page & Modal Definitions ---

class PageSpec(BaseModel):
   id: str
   route: str
   title: str
   icon: Optional[str] = None
   layout: Optional[LayoutConfig] = None
   blocks: List[BlockSpec] = []
   variables: Optional[Dict[str, Any]] = {}  # Page-level state


class ModalSpec(BaseModel):
   id: str
   title: str
   size: Literal["small", "medium", "large", "fullscreen"] = "medium"
   blocks: List[BlockSpec] = []


class NavItem(BaseModel):
   name: str
   label: str
   icon: Optional[str] = None
   route: Optional[str] = None
   children: Optional[List["NavItem"]] = None


class GlobalAction(BaseModel):
   id: str
   label: str
   icon: Optional[str] = None
   action: str
   config: Dict[str, Any] = {}


# --- Theme ---

class ThemeSpec(BaseModel):
   primaryColor: Optional[str] = "#6366f1"
   mode: Optional[Literal["light", "dark"]] = "dark"
   fontFamily: Optional[str] = None


# --- App Info ---

class AppInfo(BaseModel):
   name: str
   slug: str
   description: Optional[str] = None
   theme: Optional[ThemeSpec] = None


# --- UI Spec V3 ---

class UISpecV3(BaseModel):
   navigation: List[NavItem] = []
   pages: List[PageSpec] = []
   modals: Optional[List[ModalSpec]] = []
   globalActions: Optional[List[GlobalAction]] = []


# --- Blueprint V3 (Main Schema) ---

class BlueprintV3(BaseModel):
   version: Literal[3] = 3
   app: AppInfo
   backend: BackendConfig = BackendConfig()
   data: DataSpec
   security: SecuritySpecV3
   ui: UISpecV3


# ============================================================================
# BLUEPRINT V2 SCHEMA (Legacy - for backwards compatibility)
# ============================================================================

class PermissionRule(BaseModel):
   role: str
   resource: str
   actions: Dict[str, bool]  # list, read, create, update, delete


class FilterExpression(BaseModel):
   equals: Optional[List[str]] = None
   in_: Optional[List[str]] = Field(None, alias="in")
   and_: Optional[List["FilterExpression"]] = Field(None, alias="and")
   or_: Optional[List["FilterExpression"]] = Field(None, alias="or")

   class Config:
      populate_by_name = True


class RowFilterRule(BaseModel):
   role: str
   resource: str
   filter: FilterExpression


class SecuritySpecV2(BaseModel):
   roles: List[str]
   permissions: List[PermissionRule]
   rowFilters: Optional[List[RowFilterRule]] = []


class DataSourceSpecV2(BaseModel):
   table: str  # V2 uses "table"
   filters: Optional[List[FilterSpec]] = []
   orderBy: Optional[List[OrderSpec]] = []
   limit: Optional[int] = None
   include: Optional[List[str]] = []
   realtime: Optional[bool] = False


class BlockSpecV2(BaseModel):
   id: str
   type: str
   dataSource: Optional[DataSourceSpecV2] = None
   props: Dict[str, Any] = {}
   actions: Optional[List[ActionConfig]] = []
   gridArea: Optional[str] = None
   className: Optional[str] = None
   visibility: Optional[VisibilityRule] = None
   children: Optional[List["BlockSpecV2"]] = []


class PageSpecV2(BaseModel):
   id: str
   route: str
   title: str
   icon: Optional[str] = None
   layout: Optional[LayoutConfig] = None
   blocks: List[BlockSpecV2] = []
   variables: Optional[Dict[str, Any]] = {}


class ModalSpecV2(BaseModel):
   id: str
   title: str
   size: Literal["small", "medium", "large", "fullscreen"] = "medium"
   blocks: List[BlockSpecV2] = []


class UISpecV2(BaseModel):
   navigation: List[NavItem] = []
   pages: List[PageSpecV2] = []
   modals: Optional[List[ModalSpecV2]] = []
   globalActions: Optional[List[GlobalAction]] = []


class TableSpecV2(BaseModel):
   name: str  # snake_case for V2
   primaryKey: str = "id"
   columns: List[ColumnSpec]


class RelationshipSpecV2(BaseModel):
   type: Literal["many_to_one", "one_to_many"]
   fromTable: str
   fromColumn: str
   toTable: str
   toColumn: str
   lookupLabelColumn: Optional[str] = None


class DataSpecV2(BaseModel):
   tables: List[TableSpecV2]
   relationships: Optional[List[RelationshipSpecV2]] = []


class BlueprintV2(BaseModel):
   version: Literal[2] = 2
   app: AppInfo
   data: DataSpecV2
   security: SecuritySpecV2
   ui: UISpecV2


# ============================================================================
# Union type for all Blueprint versions
# ============================================================================

Blueprint = Union[BlueprintV3, BlueprintV2]


# ============================================================================
# Response schemas
# ============================================================================

class BlueprintResponse(BaseModel):
   id: UUID
   app_id: UUID
   version: int
   blueprint_json: Dict[str, Any]
   blueprint_hash: Optional[str]
   validation_status: ValidationStatus
   validation_errors: Optional[Dict[str, Any]]
   created_at: datetime

   class Config:
      from_attributes = True
